### PR TITLE
[ISSUE-152] Added SCRAM functionality during user registration

### PR DESCRIPTION
**Pull request by huzaifamk**
## Description:

**[MOCKED]** - Assume that iterationCount and salt were returned in the server response from register-user-precheck endpoint (check https://github.com/globe-and-citizen/layer8/issues/151 for more details), and mock it for now to unblock development of the final step in register-user flow.

**[DONE]** - In server/assets-v1/templates/src/pages/user_portal/register.html  generate storedKey and serverKey:
saltedPassword = pbkdf2(password, salt, iterationCount)
clientKey = HMAC-SHA256(saltedPassword, `Client Key`)
serverKey = HMAC-SHA256(saltedPassword, `Server Key`)
storedKey = SHA256(clientKey)
Look up about [HMAC](https://en.wikipedia.org/wiki/HMAC) and [PBKDF2](https://en.wikipedia.org/wiki/PBKDF2) here.

**[DONE]** - Send storedKey and serverKey, as well as other user specific data (username, first_name, last_name, display_name, country, public_key) to the backend as the request to /api/v1/register-user endpoint.

**[DONE]** - Modify the /api/v1/register-user endpoint:
Save user specific data, storedKey and serverKey into the users table instead of the password. We should make it via the update query, because row representing the user was already added in https://github.com/globe-and-citizen/layer8/issues/151.

**[NOT REMOVED FOR NOW AS IT WILL BREAK THE CURRENT FLOW, BUT I HAVE MADE THE CURRENT CODE TO USE /api/v2 PATH SO WE CAN CONVERT IT TO V1 ONCE MERGING IS COMPLETE WITH JAVOKHIR CODE]** - Remove other logic such as salt generation since it's already done in the precheck endpoint.

**[DONE]** - Add tests to cover the newly implemented logic, i.e. changes in controller.go, service.go and repository.go.

**[DONE]** -Add migration scripts to modify the schema of users table, i.e. add new columns storedKey and serverKey.

### SECTION 1: RESOURCE SERVER (http://localhost:5001/)
- [x] Created a new user on the Layer8 Portal

### SECTION 2: WE'VE GOT POEMS (http://localhost:5173/)
- [ ] Login as 'tester', '1234'
- [ ] Choose login with layer8 
- [ ] Logging in with the newly registered credentials from Section 1 succeeds in popup
- [ ] User chooses to share their new "Username" & "Country" from the Layer8 Resource Server
- [ ] Choose poems 1 - 3 works correctly 

### SECTION 3: IMSHARER (http://localhost:5174/)
- [ ] Uploads an image works
- [ ] Reload leads to instant reload (demonstrating proper caching)

### CODE REVIEW

File: server/resource_server/repository/repository.go [hunk 1]
```diff
@@ -69,6 +69,73 @@ func (r *Repository) RegisterUser(req dto.RegisterUserDTO, hashedPassword string
...
+	tx := r.connection.Begin()
+	if err := tx.Where("username = ?", req.Username).First(&user).Error; err != nil {
+		if err == gorm.ErrRecordNotFound {
+			tx.Rollback()
+			return fmt.Errorf("user not found")
+		}
+		tx.Rollback()
+		return err
+	}
+
```
Comment (interestIngc) [inline server/resource_server/repository/repository.go:98]: Instead of making two sql queries: an initial select query and then an update statement, we can just go with a single update:
```
	tx.Where("username = ?", req.Username).Updates(map[string]interface{}{
		"first_name": req.FirstName,
		"last_name":  req.LastName,
		// etc
	})
```
WDYT?

Comment (huzaifamk) [reply server/resource_server/repository/repository.go]: Done.

**Comment by huzaifamk** [comment]
Ready for review!

File: server/resource_server/repository/repository_test.go [hunk 2]
```diff
@@ -1206,3 +1209,165 @@ func TestUpdateUserPassword_Success(t *testing.T) {
...
+		PublicKey:   publicKey,
+		StoredKey:   storedKey,
+		ServerKey:   serverKey,
+	}
+	err = repository.RegisterUserv2(userDto)
+
+	assert.NotNil(t, err)
+	assert.Nil(t, mock.ExpectationsWereMet(), "There were unfulfilled expectations!")
+}
+
```
Comment (interestIngc) [inline server/resource_server/repository/repository_test.go:1246]: We can add one more unit test here -- when the `UPDATE "users"...` query fails, but the initial `SELECT * from "users"` succeeds.

File: server/resource_server/controller/controller_test.go [hunk 3]
```diff
@@ -1852,3 +1863,141 @@ func setMockServiceInContext(req *http.Request) *http.Request {
...
+
+	assert.Equal(t, http.StatusBadRequest, rr.Code)
+
+	response := decodeResponseBodyForErrorResponse(t, rr)
+
+	assert.False(t, response.IsSuccess)
+	assert.Equal(t, "Input json is invalid", response.Message)
+	assert.NotNil(t, response.Error)
+}
+
```
Comment (interestIngc) [inline server/resource_server/controller/controller_test.go:1963]: We can add one more test case -- when mockService returns an error != nil, assert that an error message is returned to the user with an appropriate status code.

File: server/resource_server/controller/controller_test.go [hunk 4]
```diff
@@ -183,6 +183,16 @@ func (m *MockService) UpdateUserPassword(username string, newPassword string, sa
...
 }
 
+// Mock RegisterUserPrecheck method for unit tests
+func (m *MockService) RegisterUserPrecheck(req dto.RegisterUserPrecheckDTO, iterCount int) (models.RegisterUserPrecheckResponseOutput, error) {
+	return models.RegisterUserPrecheckResponseOutput{}, nil
+}
+
+// Mock RegisterUser method for unit tests
+func (m *MockService) RegisterUserv2(req dto.RegisterUserDTOv2) error {
+	return nil
```
Comment (interestIngc) [inline server/resource_server/controller/controller_test.go:193]: We can add a custom function here to mock the behavior of `service.RegisterUserv2`, similar to how it's done in `func (m *MockService) UpdateUserPassword` (line 182)

**Review by interestIngc — COMMENTED** [review]
The PR looks great! Thank you, Huzaifa!
There are a few comments related to unit testing, overall the code looks very good!
I'll check this PR once again to give a final approval, when we merge the PR from Javokhir. Otherwise, it's in an almost ready shape to get merged (add more unit testing and then rebase onto the new version of development branch).

File: server/resource_server/utils/utils.go [hunk 5]
```diff
@@ -58,6 +60,23 @@ func SaltAndHashPassword(password string, salt string) string {
...
+	h.Write([]byte(saltedPassword))
+	return hex.EncodeToString(h.Sum(nil))
+}
+
+func GenerateSHA256Hash(clientKey string) string {
+	h := sha256.New()
+	h.Write([]byte(clientKey))
+	return hex.EncodeToString(h.Sum(nil))
+}
+
```
[the remark this answers is not in the document]
Comment (huzaifamk) [reply server/resource_server/utils/utils.go:79]: Removed.

File: server/resource_server/repository/repository_test.go (hunk 2 above)
Comment (huzaifamk) [reply server/resource_server/repository/repository_test.go:1246]: Done.

File: server/resource_server/controller/controller_test.go (hunk 3 above)
Comment (huzaifamk) [reply server/resource_server/controller/controller_test.go:1963]: Done.

File: server/resource_server/controller/controller_test.go (hunk 4 above)
Comment (huzaifamk) [reply server/resource_server/controller/controller_test.go]: Thanks for example, this was very helpful!

**Comment by huzaifamk** [comment]
Made changes as per comments, also merged latest development branch into my branch.

Ready for review again @interestIngc

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+      </div>
+      <div
+        :class="showToast ? 'opacity-100' : 'opacity-0 pointer-events-none'"
+        class="fixed top-3 right-3 bg-red-500 text-white p-2 rounded-md transition-opacity ease-in-out duration-500 z-50"
+      >
+        {{ toastMessage }}
+      </div>
+    </div>
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/crypto-js/4.0.0/crypto-js.min.js"></script>
+    <script src="https://requirejs.org/docs/release/2.3.7/minified/require.js"></script>
```
Comment (interestIngc) [inline server/assets-v1/templates/src/pages/user_portal/register_v2.html:207]: Do we need this dependency? Seems like require is not used anywhere within the script section

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+        //   console.log(
+        //     "IterCount: ",
+        //     registerPrecheckResponseBody.iterationCount
+        //   );
+
+          const dk = CryptoJS.PBKDF2(
+            registerPassword.value,
+            CryptoJS.enc.Utf8.parse(registerPrecheckResponseBody.salt),
+            {
+              keySize: 32 / 4,
```
Comment (interestIngc) [inline server/assets-v1/templates/src/pages/user_portal/register_v2.html:283]: Why do we use `keySize = 8` here? I think this means, that size of the resulting `derivedKey` is 8 bits. But my understanding is that size of the PBKDF2 output should be a multiple of the underlying hash output size, which is 160 bits in our case (since we use SHA1). This way, we should target for 160 or 320 bits as the keySize. If we set it to 8, this would make no PRF included into the final concatenation (`8 / 160 = 0`), and that would apparently result into an empty derivedKey.

Correct me if I'm wrong!

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+              iterations: registerPrecheckResponseBody.iterationCount,
+              hasher: CryptoJS.algo.SHA1,
+            }
+          );
+
+          const saltedPassword = dk.toString(CryptoJS.enc.Hex);
+
+          // Hardcoded Client and Server Keys
+          const CLIENT_KEY_ENV = "TEST_CLIENT_KEY";
+          const SERVER_KEY_ENV = "TEST_SERVER_KEY";
```
Comment (interestIngc) [inline server/assets-v1/templates/src/pages/user_portal/register_v2.html:293]: We can put it as just clientKey and serverKey

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+
+          // HMAC for clientKey
+          const clientKeyHmac = CryptoJS.HmacSHA256(
+            CryptoJS.enc.Utf8.parse(saltedPassword),
+            CryptoJS.enc.Utf8.parse(CLIENT_KEY_ENV)
+          );
+          const clientKey = clientKeyHmac.toString(CryptoJS.enc.Hex);
```
Comment (interestIngc) [inline server/assets-v1/templates/src/pages/user_portal/register_v2.html:300]: We can inline this into a single `const clientKey` (because `clientKeyHmac` is not used afterwards)

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+
+          // HMAC for serverKey
+          const serverKeyHmac = CryptoJS.HmacSHA256(
+            CryptoJS.enc.Utf8.parse(saltedPassword),
+            CryptoJS.enc.Utf8.parse(SERVER_KEY_ENV)
+          );
+          const serverKey = serverKeyHmac.toString(CryptoJS.enc.Hex);
```
Comment (interestIngc) [inline server/assets-v1/templates/src/pages/user_portal/register_v2.html:307]: Same for `const serverKey `

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html [hunk 11]
```diff
@@ -0,0 +1,380 @@
...
+
+          // SHA256 for storedKey
+          const storedKeyHash = CryptoJS.SHA256(
+            CryptoJS.enc.Utf8.parse(clientKey)
+          );
+          const storedKey = storedKeyHash.toString(CryptoJS.enc.Hex);
```
Comment (interestIngc) [inline server/assets-v1/templates/src/pages/user_portal/register_v2.html:313]: The same for `const storedKey` :)

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+
+        //   console.log("serverKey: ", serverKey);
+        //   console.log("storedKey: ", storedKey);
```
Comment (interestIngc) [inline server/assets-v1/templates/src/pages/user_portal/register_v2.html:316]: We should remove this comments before the PR gets merged!

**Comment by interestIngc** [comment]
Great! I've just left a few more comments for the frontend logic. 
Thanks for a detailed description of the PR!

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html (hunk 6 above)
Comment (huzaifamk) [reply server/assets-v1/templates/src/pages/user_portal/register_v2.html:207]: Yes, no need of that, I was trying something else in the start, for importing the crypto package, but then tired other way so this is no longer needed, removing it :heavy_check_mark:

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html (hunk 7 above)
Comment (huzaifamk) [reply server/assets-v1/templates/src/pages/user_portal/register_v2.html:283]: I searched about it, and the calculation for it is as follows: 

The reason keySize is given as 32 / 4 instead of just 8 is that the keySize parameter in CryptoJS's PBKDF2 function expects the size of the key in words, not bytes. In CryptoJS, a word is 32 bits (4 bytes). So, when you specify 32 / 4, you're actually specifying a key size of 8 bytes (since 32 bits / 4 bytes per word = 8 bytes). If you were to write 8 directly, CryptoJS would interpret it as 8 words, which would be equivalent to 32 bytes (since 8 words * 4 bytes per word = 32 bytes). So, in this case, 32 / 4 and 8 are equivalent, but the former is more explicit about the unit of measurement (words vs bytes).

I will change this to keySize: 40 / 8 which will result in a key size of 160 bits.

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html (hunk 12 above)
Comment (huzaifamk) [reply server/assets-v1/templates/src/pages/user_portal/register_v2.html:316]: Removed.

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html (hunk 9 above)
Comment (huzaifamk) [reply server/assets-v1/templates/src/pages/user_portal/register_v2.html]: Done.

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html (hunk 8 above)
Comment (huzaifamk) [reply server/assets-v1/templates/src/pages/user_portal/register_v2.html]: Done, added as clientKeySecret and serverKeySecret since `clientKey and serverKey` are already being used on line 300 and afterwards.

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html (hunk 10 above)
Comment (huzaifamk) [reply server/assets-v1/templates/src/pages/user_portal/register_v2.html]: Done.

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html (hunk 11 above)
Comment (huzaifamk) [reply server/assets-v1/templates/src/pages/user_portal/register_v2.html]: Done.

**Comment by huzaifamk** [comment]
Made changes in the last meeting according to discussion, ready for final review @interestIngc

File: migrations/000006_alter_users_table.down.sql [hunk 13]
```diff
@@ -0,0 +1,3 @@
+ALTER TABLE IF EXISTS users
```
[the remark this answers is not in the document]
Comment (huzaifamk) [reply migrations/000006_alter_users_table.down.sql:1]: Done.

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html [hunk 14]
```diff
@@ -189,46 +219,85 @@ <h3 class="modal__header">
...
+            return;
+          }
 
-          const registerPrecheckResponseBody = await respOne.json();
-          console.log("Response from Register Precheck", registerPrecheckResponseBody)
-          return;
-          // From here implementation should be continued
+          const dk = CryptoJS.PBKDF2(
+            registerPassword.value,
+            CryptoJS.enc.Utf8.parse(registerPrecheckResponseBody.salt),
```
Comment (interestIngc) [inline server/assets-v1/templates/src/pages/user_portal/register_v2.html:272]: Here we should use `registerPrecheckResponseBody.data.salt` instead of `registerPrecheckResponseBody.salt` since the response from the `register-precheck` endpoint comes in the form of the `Response` struct

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html [hunk 15]
```diff
@@ -189,46 +219,85 @@ <h3 class="modal__header">
...
+            {
+              // TODO: To explore more about this
+              keySize: 160 / 32,
+              iterations: registerPrecheckResponseBody.iterationCount,
```
Comment (interestIngc) [inline server/assets-v1/templates/src/pages/user_portal/register_v2.html:276]: Same here, please change it to `registerPrecheckResponseBody.data.iterationCount`

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html (hunk 14 above)
Comment (huzaifamk) [reply server/assets-v1/templates/src/pages/user_portal/register_v2.html:272]: Done.

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html (hunk 15 above)
Comment (huzaifamk) [reply server/assets-v1/templates/src/pages/user_portal/register_v2.html]: Done.

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html [hunk 16]
```diff
@@ -189,46 +219,85 @@ <h3 class="modal__header">
...
-          const resp = await window.fetch("[[ .ProxyURL ]]/api/v1/register-user", {
+          // HMAC for clientKey
+          const clientKey = CryptoJS.HmacSHA256(
+            CryptoJS.enc.Hex.parse(saltedPassword),
+            CryptoJS.enc.Utf8.parse("Client Key")
+          ).toString(CryptoJS.enc.Hex);
+
+          // HMAC for serverKey
+          const serverKey = CryptoJS.HmacSHA256(
+            CryptoJS.enc.Utf8.parse(saltedPassword),
```
Comment (interestIngc) [inline server/assets-v1/templates/src/pages/user_portal/register_v2.html:291]: Here we can change it to Hex as well

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html [hunk 17]
```diff
@@ -189,46 +219,85 @@ <h3 class="modal__header">
...
+            return;
+          }
 
-          const registerPrecheckResponseBody = await respOne.json();
-          console.log("Response from Register Precheck", registerPrecheckResponseBody)
-          return;
-          // From here implementation should be continued
+          const dk = CryptoJS.PBKDF2(
+            registerPassword.value,
+            CryptoJS.enc.Utf8.parse(registerPrecheckResponseBody.data.salt),
```
[the remark this answers is not in the document]
Comment (huzaifamk) [reply server/assets-v1/templates/src/pages/user_portal/register_v2.html:272]: Agreed, done.

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html [hunk 18]
```diff
@@ -189,46 +219,85 @@ <h3 class="modal__header">
...
+          );
 
-          currMnemonic.value = mnemonic.generateBip39Mnemonic();
-          const keyPair = mnemonic.getPrivateAndPublicKeys(currMnemonic.value);
+          const saltedPassword = dk.toString(CryptoJS.enc.Hex);
 
-          const resp = await window.fetch("[[ .ProxyURL ]]/api/v1/register-user", {
+          // HMAC for clientKey
+          const clientKey = CryptoJS.HmacSHA256(
+            CryptoJS.enc.Utf8.parse(saltedPassword),
```
[the remark this answers is not in the document]
Comment (huzaifamk) [reply server/assets-v1/templates/src/pages/user_portal/register_v2.html]: Done.

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html (hunk 16 above)
Comment (huzaifamk) [reply server/assets-v1/templates/src/pages/user_portal/register_v2.html]: Done.

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html [hunk 19]
```diff
@@ -189,46 +219,85 @@ <h3 class="modal__header">
...
+              // TODO: To explore more about this
+              keySize: 160 / 32,
+              iterations: registerPrecheckResponseBody.data.iterationCount,
+              hasher: CryptoJS.algo.SHA1,
+            }
+          );
 
-          currMnemonic.value = mnemonic.generateBip39Mnemonic();
-          const keyPair = mnemonic.getPrivateAndPublicKeys(currMnemonic.value);
+          const saltedPassword = dk.toString(CryptoJS.enc.Hex);
```
Comment (interestIngc) [inline server/assets-v1/templates/src/pages/user_portal/register_v2.html:281]: We could just inline variable `dk` into `saltedPassword` (since we don't use it separately afterwards)

**Review by interestIngc — APPROVED** [review]

File: server/assets-v1/templates/src/pages/user_portal/register_v2.html (hunk 19 above)
Comment (huzaifamk) [reply server/assets-v1/templates/src/pages/user_portal/register_v2.html:281]: Done.
